backend/server.go

```diff
@@ -104,34 +137,31 @@ const MethodPost = "POST"
 func DefineRoutes() http.Handler {
 	mux := pat.New()
 
-	// static files
-	// {
-	// 	staticDirectoryName := "static"
-	// 	staticDirectoryPaddedWithSlashes := "/" + staticDirectoryName + "/"
-	//
-	// 	fileServer := http.FileServer(http.Dir(staticDirectoryName))
-	//
-	// 	mux.Handle(
-	// 		staticDirectoryPaddedWithSlashes,
-	// 		http.StripPrefix(staticDirectoryPaddedWithSlashes, fileServer))
-	// }
-
 	// Redirects
 
-	mux.Post("/api/games/{gameId}/role_pool", WrapGameApiEndpoint(AssignRolePool))
+	mux.Put("/api/games/{gameId}/role_pool", WrapGameApiEndpoint(AssignRolePool))
 	mux.Get("/api/games/{gameId}/role_pool", WrapGameApiEndpoint(GetRolePool))
 
 	mux.Post("/api/games/{gameId}/start", WrapGameApiEndpoint(StartGame))
 
-	// mux.Put("/api/games/{gameId}/players/{playerId}/action", WrapPlayerApiEndpoint(GetPlayerInfo))
 	mux.Get("/api/games/{gameId}/players/{playerId}", WrapPlayerApiEndpoint(GetPlayerInfo))
 	mux.Post("/api/games/{gameId}/players", WrapGameApiEndpoint(CreatePlayer))
 
 	mux.Post("/api/games", WrapApiEndpoint(CreateGame))
 
+	mux.Post("/api/games/{gameId}/player/{playerId}/do_action", WrapPlayerApiEndpoint(DoAction))
+
+	mux.PathPrefix("/").Methods(http.MethodOptions).HandlerFunc(AllowAllCors)
```

Comment:
why do we need to do this and allowCors in endpoint API?

Reply:
cors works differently for GET than other methods. Since GET shouldn't have any side effects, the browser will issue a GET to other origins. Just if the response doesn't have the proper headers the browser won't let the application see the response.

For most POSTs and all PUT, DELETE and PATCH the browser won't even send the request to another origin. Instead it issues a preflight request of method type OPTIONS which is basically the browser asking for permission to send the real request. If the server doesn't respond at all or doesn't respond with the right headers the browser won't issue the real request. (This is dumb imo. It should be the server's responsibility to not execute side effects from untrusted origins)

Line 154 here is to handle preflight requests. The `allowCors` is to set the headers on GET requests.
See
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
and
https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

Reply:
This is some crazy pants stuff.